cpp/dolfinx/fem/discreteoperators.h

```diff
@@ -195,149 +200,185 @@ void interpolation_matrix(const FunctionSpace& V0, const FunctionSpace& V1,
-  std::vector<T> A(space_dim0 * space_dim1);
+  std::vector<double> basis_values_b(Xshape[0] * bs0 * dim0 * e1->value_size());
+  mdspan3_t basis_values(basis_values_b.data(), Xshape[0], bs0 * dim0,
+                         e1->value_size());
+  std::vector<double> mapped_values_b(Xshape[0] * bs0 * dim0
+                                      * e1->value_size());
+  mdspan3_t mapped_values(mapped_values_b.data(), Xshape[0], bs0 * dim0,
+                          e1->value_size());
+
+  auto pull_back_fn1 = e1->basix_element().map_fn<u_t, U_t, K_t, J_t>();
+
+  std::vector<double> coordinate_dofs_b(num_dofs_g * 3);
+  stdex::mdspan<double, stdex::extents<std::size_t, stdex::dynamic_extent, 3>>
+      coordinate_dofs(coordinate_dofs_b.data(), num_dofs_g, 3);
+  std::vector<double> basis0_b(Xshape[0] * dim0 * value_size0);
+  mdspan3_t basis0(basis0_b.data(), Xshape[0], dim0, value_size0);
+
+  // Buffers
+  std::vector<T> Ab(space_dim0 * space_dim1);
   std::vector<T> local1(space_dim1);
 
-  std::vector<std::size_t> shape
-      = {X.shape(0), (std::size_t)element1->value_size(), space_dim0};
-  auto _A = xt::adapt(A, shape);
-
   // Iterate over mesh and interpolate on each cell
   auto cell_map = mesh->topology().index_map(tdim);
   assert(cell_map);
   std::int32_t num_cells = cell_map->size_local();
-  auto _coordinate_dofs
-      = xt::view(coordinate_dofs, xt::all(), xt::xrange(0, gdim));
+  auto coord_dofs = stdex::submdspan(coordinate_dofs, stdex::full_extent,
+                                     std::pair(0, gdim));
+
   for (std::int32_t c = 0; c < num_cells; ++c)
   {
     // Get cell geometry (coordinate dofs)
     auto x_dofs = x_dofmap.links(c);
     for (std::size_t i = 0; i < x_dofs.size(); ++i)
     {
       common::impl::copy_N<3>(std::next(x_g.begin(), 3 * x_dofs[i]),
-                              std::next(coordinate_dofs.begin(), 3 * i));
+                              std::next(coordinate_dofs_b.begin(), 3 * i));
     }
 
     // Compute Jacobians and reference points for current cell
-    J.fill(0);
-    for (std::size_t p = 0; p < X.shape(0); ++p)
+    std::fill(J_b.begin(), J_b.end(), 0);
+    for (std::size_t p = 0; p < Xshape[0]; ++p)
     {
-      auto _J = xt::view(J, p, xt::all(), xt::all());
-      cmap.compute_jacobian(dphi, _coordinate_dofs, _J);
-      cmap.compute_jacobian_inverse(_J, xt::view(K, p, xt::all(), xt::all()));
-      detJ[p] = cmap.compute_jacobian_determinant(_J);
+      auto _J = stdex::submdspan(J, p, stdex::full_extent, stdex::full_extent);
+      cmap.compute_jacobian(dphi, coord_dofs, _J);
+      auto _K = stdex::submdspan(K, p, stdex::full_extent, stdex::full_extent);
+      cmap.compute_jacobian_inverse(_J, _K);
+      detJ[p] = cmap.compute_jacobian_determinant(_J, det_scratch);
     }
 
-    // Get evaluated basis on reference, apply DOF transformations, and
+    // Copy evaluated basis on reference, apply DOF transformations, and
     // push forward to physical element
```

Comment:
Could we add a note here saying that we should not use `basis_reference0`, i.e. that this is a bug.